### PR TITLE
Simplify grammar rules around ASYNC

### PR DESCRIPTION
**Pull request by gvanrossum**
Inspired by some criticism on python-dev I looked into whether we can simplify the grammar by writing e.g. (omitting actions and item names):
```
  function_def_raw:
      | [ASYNC] 'def' NAME '(' parameters? ')' ['->' annotation] ':' block
```
instead of
```
  function_def_raw:
      | ASYNC 'def' NAME '(' parameters? ')' ['->' annotation] ':' block
      | 'def' NAME '(' parameters? ')' ['->' annotation] ':' block
```
This seems to work just fine. I doubt we could even measure whether it's slower or faster, but I suspect that it's a tad faster, and less code to boot!

(**UPDATE:**) The actions are the same except for the function being called, and you can use a C conditional expression to choose which function to call. I did this for `async def`, `async for` and `async with`. I also tweaked one occurrence of `async for` that was already collapsed to write the test in less code.

### CODE REVIEW

File: Grammar/python.gram [hunk 1]
```diff
@@ -142,20 +142,14 @@ while_stmt[stmt_ty]:
     | 'while' a=named_expression ':' b=block c=[else_block] { _Py_While(a, b, c, EXTRA) }
 
 for_stmt[stmt_ty]:
-    | ASYNC 'for' t=star_targets 'in' ex=star_expressions ':' b=block el=[else_block] {
-        _Py_AsyncFor(t, ex, b, el, NULL, EXTRA) }
-    | 'for' t=star_targets 'in' ex=star_expressions ':' b=block el=[else_block] {
-        _Py_For(t, ex, b, el, NULL, EXTRA) }
+    | is_async=[ASYNC] 'for' t=star_targets 'in' ex=star_expressions ':' b=block el=[else_block] {
+        (is_async ? _Py_AsyncFor : _Py_For)(t, ex, b, el, NULL, EXTRA) }
```
Comment (lysnikolaou) [inline Grammar/python.gram:146]: Wow! Didn't even know that this was possible. Then again, function names are just pointers, so it's logical in hindsight.

**Comment by lysnikolaou** [comment]
> This seems to work just fine. I doubt we could even measure whether it's slower or faster, but I suspect that it's a tad faster, and less code to boot!

I'd suspect that it's tiny bit slower, since a new function is now called for the optional `ASYNC`, which wasn't called before and everything was happening in `function_def_raw_rule`. Would that make sense?

**Comment by gvanrossum** [comment]
Looking at diff for the generated code I don't see an extra function being called. Or maybe I'm missing something? (The diff is a little confusing.)

**Comment by lysnikolaou** [comment]
My bad, it seems the generator does not generate "temporary" rules for single-element optionals and I though it did. Sorry!

**Pull request merged**